osu.Game/Rulesets/RulesetStore.cs

```diff
@@ -158,19 +150,35 @@ private static void loadRulesetFromFile(string file)
             }
         }
 
-        private static void addRuleset(Assembly assembly)
+        private void addRuleset(Assembly assembly)
         {
-            if (loaded_assemblies.ContainsKey(assembly))
+            if (loadedAssemblies.ContainsKey(assembly))
                 return;
 
             try
             {
-                loaded_assemblies[assembly] = assembly.GetTypes().First(t => t.IsPublic && t.IsSubclassOf(typeof(Ruleset)));
+                loadedAssemblies[assembly] = assembly.GetTypes().First(t => t.IsPublic && t.IsSubclassOf(typeof(Ruleset)));
             }
             catch (Exception e)
             {
                 Logger.Error(e, $"Failed to add ruleset {assembly}");
             }
         }
+
+        ~RulesetStore()
+        {
+            Dispose(false);
+        }
+
+        public void Dispose()
+        {
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            AppDomain.CurrentDomain.AssemblyResolve -= resolveRulesetAssembly;
+        }
```

Comment:
This `Dispose()` implementation seems to go against what I thought I knew about the dispose pattern, namely that:

* finalizers should be used pretty much only when there are managed resources to dispose, and
* disposing managed resources inside `Dispose(bool disposing)` should be guarded by the `disposing` parameter.

As far as I understand this could attempt to unsubscribe from `AssemblyResolve` twice (I do however know [finalizers are not guaranteed to run in general](https://ericlippert.com/2015/05/18/when-everything-you-know-is-wrong-part-one/)). I *could* very well be full of it, though (I'd even say it's likely).

Reply:
Oh, and [part two of that post](https://ericlippert.com/2015/05/21/when-everything-you-know-is-wrong-part-two/) says that apparently it might be wrong to even assume the `AssemblyResolve` event was even subscribed to in the first place, since the ctor didn't necessarily fully execute...

Reply:
Unsubscribing from events that haven't been subscribed to has no ill effects.

It's sometimes fine to use the finalizer in a managed context (e.g. `ArrayPool<>`). So long as you understand the implications of doing so (thread safety, etc). In this case the usage is fine, but it won't actually do anything since a managed reference to the object is held from an external source if `Dispose()` is not invoked, so I'll remove it.